services/QuillLMS/spec/support/shared_contexts/snapshots/snapshots_cte.rb

```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+RSpec.shared_context 'Snapshots CTE' do
+  # BigQuery doesn't support CTE aliases with period in the name.
+  # So, we need to remove the 'lms.' and 'special.' prefixes from 'FROM' and 'JOIN' clauses
+  let(:query) { pg_query.gsub(/(FROM|JOIN)\s+(?:lms|special)\.(\w+)/, '\1 \2') }
+
+  let(:bq_query) do
+    <<-SQL
+      WITH
+        #{cte}
+      #{query}
+    SQL
+  end
+
+  let(:results) { QuillBigQuery::Runner.execute(bq_query) }
+
+  def cte
```

Comment:
Is there a utility file we can stash these functions in?  And then maybe replace the `let(:bq_query)` block with a function call?  Something like `big_query_with_context(query, array_of_relevant_models)`?  Then you can do `let(:query) { big_query_with_context(described_class.new(params).query_to_run, [classrooms, users, ...]) }`  This will make it easier to utilize this outside of the Snapshots shared context (such as progress reports).

Reply:
Good idea.  I will update.